util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+# Based on the blog post "How to cluster images based on visual similarity" 
+# written by Gabe Flomo.
+
+
+# for loading/processing the images  
+from keras.preprocessing.image import load_img 
+from keras.preprocessing.image import img_to_array 
+from keras.applications.vgg16 import preprocess_input 
+
+# models 
+from keras.applications.vgg16 import VGG16 
+from keras.models import Model
+
+# clustering and dimension reduction
+from sklearn.cluster import KMeans
+from sklearn.decomposition import PCA
+
+# for everything else
+import argparse
+import os
+import numpy as np
+import matplotlib.pyplot as plt
+from PIL import Image
+from tqdm import tqdm
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Clustering')
+    if 'SM_OUTPUT_DATA_DIR' in os.environ: 
+
+        parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])
+        #parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+        parser.add_argument('--input-data-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+    else: 
+        parser.add_argument('--output-data-dir', type=str)
+        parser.add_argument('--input-data-dir', type=str)   
+    
+    parser.add_argument('--analyze-clusters', type=int, default=3)   
+
+    args = parser.parse_args()
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
+
+    detection_train_images = []
+    # creates a ScandirIterator aliased as files
+    print('------------ getting image names ------------')
+    with os.scandir(output_path) as files:
+      # loops through each file in the directory
+        for file in files:
+            if file.name.endswith('.jpg'):
+              # adds only the image files to the detection_train_imagess list
+                detection_train_images.append(file.name)
+   
+    # print('get bucket info')
+    # bucket = "calvinandpogs-ee148"
+    # detection_train_images_path = "atrw/detection/train"
+    # s3 = boto3.resource("s3")
+    # s3_bucket = s3.Bucket(bucket)
+    # print('bucket info collected')
+
+    # # this list holds all the image filename
+    # detection_train_images = [f.key.split(detection_train_images_path + "/")[1] for f in s3_bucket.objects.filter(Prefix=detection_train_images_path).all()]
+    print(str(len(detection_train_images)) + ' images collected')
+
+        
+    model = VGG16()
+    model = Model(inputs = model.inputs, outputs = model.layers[-2].output)
+
+    def extract_features(file, model):
```

Comment:
This does not need to be a private inner-function. Just define it as a normal function outside of main()

Reply:
in-fact it really should not be an inner function. The variable names in this scope are conflicting and can cause logical errors for our variable references.